File: cpp/problem788.cpp
#include "iostream"
using namespace std;


class Solution {
public:
    int rotatedDigits(int N) {
        int count = 0, rotated;
        for (int i = 1; i <= N; ++i)
            if( rotated = rotate(i) ) count += rotated != i;
        return count;
    }
    int rotate(int n){
        int rotated = 0, base = 1;
        while( n ){
            int rem = n%10;
            switch(rem){
                case 0: case 1: case 8:
                    rotated += base * rem;
                    break;
                case 2:
                    rotated += base * 5;
                    break;
                case 5:
                    rotated += base * 2;
                    break;
                case 6:
                    rotated += base * 9;
                    break;
                case 9:
                    rotated += base * 6;
                    break;
                default:
                    return 0;
            }
            base *= 10;
            n /= 10;
        }
        return rotated;
    }
};


int main(int argc, char const *argv[]){
    Solution sol;
    cout << sol.rotatedDigits( 10 ) << endl;
    return 0;
}

File: python/problem788.py
import time
import re
class Test:
    def test(self,s,repeat):
        t1=time.time()
        for i in range(repeat):
            s.rotatedDigits(N)
        t2=time.time()
        print('time=',t2-t1)

class Solution:
    def rotatedDigits(self, N):
        """
        :type N: int
        :rtype: int
        """
        set1={3,4,7}
        set2={0,1,8}
        cnt=0
        for i in range(1,N+1):
            if any(int(k) in set1 for k in str(i)) or all(int(k) in set2 for k in str(i)):
                cnt+=1
        return N-cnt
        
# https://leetcode.com/problems/rotated-digits/discuss/174242/Python-re               
class Solution2(object):
    def rotatedDigits(self, N):
        """
        :type A: List[int]
        :type B: List[int]
        :rtype: List[int]
        """
        count = 0
        for i in range(1,N+1):
            i = str(i)
            if re.search('[347]+', i) or re.search('[018]+',i) and not re.search('[2569]', i):
                count += 1
        return N-count


                  
                
                
if __name__ == '__main__':
    N=1008950
    s=Solution()
    s1=Solution2()
    t=Test()
    t.test(s,5)
    t.test(s1,5)